src/datasets/data_files.py

```diff
@@ -42,8 +43,12 @@ class EmptyDatasetError(FileNotFoundError):
     Split.TEST: ["test", "testing", "eval", "evaluation"],
 }
 NON_WORDS_CHARS = "-._ 0-9"
-KEYWORDS_IN_FILENAME_BASE_PATTERNS = ["**[{sep}/]{keyword}[{sep}]*", "{keyword}[{sep}]*"]
-KEYWORDS_IN_DIR_NAME_BASE_PATTERNS = ["{keyword}[{sep}/]**", "**[{sep}/]{keyword}[{sep}/]**"]
+if config.FSSPEC_VERSION < version.parse("2023.9.0"):
+    KEYWORDS_IN_FILENAME_BASE_PATTERNS = ["**[{sep}/]{keyword}[{sep}]*", "{keyword}[{sep}]*"]
+    KEYWORDS_IN_DIR_NAME_BASE_PATTERNS = ["{keyword}[{sep}/]**", "**[{sep}/]{keyword}[{sep}/]**"]
+else:
+    KEYWORDS_IN_FILENAME_BASE_PATTERNS = ["**/*[{sep}/]{keyword}[{sep}]*", "{keyword}[{sep}]*"]
+    KEYWORDS_IN_DIR_NAME_BASE_PATTERNS = ["{keyword}[{sep}/]**", "**/*[{sep}/]{keyword}[{sep}/]**"]
```

Comment:
With the current behavior (double asterisk only between slashes meaning zero or more path segments) I find confusing several patterns here:
- why do we add the slash to the sep characters here: `[{sep}/]`? I think it would be clearer to remove it as we already prefix the pattern with `**/`
- as they are:
  - the first pattern matches `dir/train.csv`, `my-train.csv` and `dir/my-train.csv`, 
  - whereas the second pattern only matches `train.csv`
- I think it would be clearer to use these patterns: `["**/*[{sep}]{keyword}[{sep}]*", "**/{keyword}[{sep}]*"]`, so that
  - the first pattern matches: `my-train.csv` and `dir/my-train.csv`
  - and the second: `train.csv` and `dir/train.csv`

Reply:
I see... they introduced new bugs... :confused: